src/Kestrel.Core/Internal/Http/HttpHeaders.cs

```diff
@@ -277,170 +277,137 @@ public static void ValidateHeaderNameCharacters(string headerCharacters)
             }
         }
 
-        public static unsafe ConnectionOptions ParseConnection(in StringValues connection)
+        public static ConnectionOptions ParseConnection(in StringValues connection)
         {
             var connectionOptions = ConnectionOptions.None;
 
             var connectionCount = connection.Count;
             for (var i = 0; i < connectionCount; i++)
             {
-                var value = connection[i];
-                fixed (char* ptr = value)
+                var value = connection[i].AsSpan();
+                var currentPosition = 0;
+                while (currentPosition < value.Length)
                 {
-                    var ch = ptr;
-                    var tokenEnd = ch;
-                    var end = ch + value.Length;
-
-                    while (ch < end)
+                    var token = GetToken(value, currentPosition);
+                    currentPosition += token.Length;
+                    var offset = 0;
+                    if (token.Length >= 9 && (token[offset] | 0x20) == 'k')
                     {
-                        while (tokenEnd < end && *tokenEnd != ',')
-                        {
-                            tokenEnd++;
-                        }
-
-                        while (ch < tokenEnd && *ch == ' ')
+                        if ((token[++offset] | 0x20) == 'e' &&
```

Comment:
This will introduce bounds-checks, because the JIT can't assume that `offset` is within the allowed bounds of `token`.

Other places as well.